bin/test_common.c

```diff
@@ -1697,18 +1724,125 @@ send_packets_one_by_one (const struct lsquic_out_spec *specs, unsigned count)
     }
 }
 
+#if HAVE_GSO
+/* UDP GSO
+ * Refs: 
+ * https://lwn.net/Articles/752956/
+ * http://vger.kernel.org/lpc_net2018_talks/willemdebruijn-lpc2018-udpgso-paper-DRAFT-1.pdf
+ */
+
+static int send_iovecs_gso(struct lsquic_out_spec *spec, 
+                           struct iovec *vecs, unsigned vcnt)
+{
+    spec->iov    = vecs;
+    spec->iovlen = vcnt;
+    LSQ_DEBUG("GSO with burst:%d", vcnt);
+    return send_packets_one_by_one (spec, 1, vcnt > 1? vecs[0].iov_len: 0);
+}
+
+/* return true if specs match */
+int match_spec(const struct lsquic_out_spec *s1,
+               const struct lsquic_out_spec *s2)
+{
+    if (s1->local_sa != s2->local_sa ||
+        s1->dest_sa  != s2->dest_sa  ||
+        s1->peer_ctx != s2->peer_ctx ||
+        s1->ecn      != s2->ecn)
+        return 0;
+    return 1;
+}
+
+/*
+ * To use GSO the flow here is:
+ * a. Check all the equal length iovs and batch them in single iovec array and
+ *    pass it in spec.
+ * b. Batching inter-spec iovs can happen only if all the spec parameters
+ *    (peer_ctx, sa, ecn) match.
+ * c. The last packet in the iovec array can be of smaller length then all the
+ *    earlier packets.
+ */
+static int
+send_packets_using_gso (const unsigned burst,
+                        const struct lsquic_out_spec *specs,
+                        unsigned count)
+{
+    struct iovec vecs[burst];
+    struct iovec *inv;
+    struct lsquic_out_spec newspec;
+    unsigned i, j, vcnt = 0;
+    int ret;
+
+    if (count == 0 || specs[0].iovlen == 0) {
+        LSQ_ERROR("Sanity failed. count=%d, specs iovlen=%zu",
+                count, specs[0].iovlen);
+        return 0;
+    }
+
+    /* Coalesce packets with same lengths, except that the last packet can be
+     * of smaller length. Colasece max gso_burst packets. */
+    for (i = 0; i < count; ++i)
+    {
+        if (vcnt == 0) {
+            memcpy(&newspec, &specs[i], sizeof(struct lsquic_out_spec));
+        } else if(!match_spec(&newspec, &specs[i])) {
+            // new specs dont match prev ones, so send the previous iovec batch
```

Comment:
When there are two connections (or more) that send packets, the lsquic engine interleaves the packets.  In that scenario, you'd end up with sending single packets most of the time.  I believe a better approach is to perform the check whether all the specs match in `sport_packets_out()`.  If they don't match, use non-GSO sending method.

Reply:
Yep, I was aware that two connections packets could be interleaved in the same call. 

The `match_spec(&newspec, &specs[i])` takes care of identifying it. If the spec differs then the batching is stopped and whatever batch is currently available is sent. Does that make sense?

Reply:
Yes, it makes sense.  But I wanted to emphasize that when there is more than one connection, one is likely to end up with specs _never_ matching and thus, given _N_ packets to send, we'll send _N_ batches of one.  That's why I suggested to perform the check earlier and use the result to pick the sending function: GSO or non-GSO.